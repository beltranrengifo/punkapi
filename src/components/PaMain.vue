<template>
  <main class="pa-main">
    <pa-grid v-if="beers && beers.length" :items="beers" />
    <pa-error v-else-if="error" :error="error" />
    <pa-no-results
      v-else-if="!beers.length"
      text="No beers around :("
      subtext="Try again!"
    />
    <pa-modal-beer />
  </main>
</template>

<script>
import { mapActions, mapGetters } from 'vuex'
import PaGrid from '@/components/PaGrid'
import PaError from '@/components/PaError'
import PaNoResults from '@/components/PaNoResults'
import PaModalBeer from '@/components/PaModalBeer'

export default {
  name: 'PaMain',

  components: {
    PaGrid,
    PaError,
    PaNoResults,
    PaModalBeer,
  },

  async created() {
    await this.fetchBeers()
  },

  computed: {
    ...mapGetters({
      beers: 'punkapi/getBeers',
      error: 'punkapi/getError',
    }),
  },

  methods: {
    ...mapActions({
      fetchBeers: 'punkapi/fetchBeers',
    }),
  },
}
</script>

<style lang="scss" scoped>
.pa-main {
  padding: rem(48) 0 rem(124);
}
</style>
