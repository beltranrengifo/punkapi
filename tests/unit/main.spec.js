import Vue from 'vue'
import Vuetify from 'vuetify'
import Vuex from 'vuex'

import { shallowMount, createLocalVue } from '@vue/test-utils'

import PaMain from '@/components/PaMain'

const localVue = createLocalVue()
Vue.use(Vuetify)
localVue.use(Vuex)

describe('Main component', () => {
  let actions
  let getters
  let store
  let wrapper

  beforeEach(() => {
    actions = { 'punkapi/fetchBeers': jest.fn() }
    getters = {
      'punkapi/getBeers': jest.fn(),
      'punkapi/getError': jest.fn(),
   }
    store = new Vuex.Store({
      state: {
        common: {
          DEFAULT_TITLE: 'PunkApi',
        },
      },
      actions,
      getters,
    })
    wrapper = shallowMount(PaMain, {
      localVue,
      store,
      computed: {
        beers: {
          get() {
            return [{ id: 1 }, { id: 2 }]
          },
          set(val) {
            return val
          }
        },
      },
    })
  })

  it('Main works and contains proper tag', () => {
    const main = wrapper.find('main')
    expect(main.exists()).toBe(true)
  })

  it('Main should match the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('Main dispatches fetchBeers when instance is created', () => {
    expect(actions['punkapi/fetchBeers']).toHaveBeenCalled()
  })

  it('Main should render no-results component if no beers in array', () => {
    const wrapper = shallowMount(PaMain, {
      localVue,
      store,
      computed: {
        beers: {
          get() {
            return []
          },
          set(val) {
            return val
          }
        },
      },
    })

    const c = wrapper.findComponent({name: 'pa-no-results'})
    expect(c.exists()).toBe(true)
  })
  it('Main should render error component if error', () => {
    const wrapper = shallowMount(PaMain, {
      localVue,
      store,
      computed: {
        beers: {
          get() {
            return []
          },
          set(val) {
            return val
          }
        },
        error: {
          get() {
            return new Error('Hi!')
          },
          set(val) {
            return val
          }
        },
      },
    })

    const c = wrapper.findComponent({name: 'pa-error'})
    expect(c.exists()).toBe(true)
  })
})
