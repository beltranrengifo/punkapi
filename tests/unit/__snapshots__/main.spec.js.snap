// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Main should match the snapshot 1`] = `
<main
  class="pa-main"
>
  <pa-grid-stub
    items="[object Object],[object Object]"
  />
</main>
`;
